feat: allow prefix for any generative model

borisdayma opened the pull request:
Allow `padding_text` for any generative model.

### Motivations

* some models require some text at the beginning such as GPT-2 which should always have a white space. GPT-2 model could now have `config.padding_text` set to `" "` to ensure this is done
* this argument adds more customization to huggingface inference widget (related issue #5553). For huggingtweets models I can set `model.config.task_specific_params['text-generation']['padding_text'] = '<|endoftext|> '`

See [example](https://gist.github.com/borisdayma/89aaf1587390340add44c3c7081bffcd) of use of `padding_text` to create negative/positive sentences.

*Side note*: this is not completely obvious that a leading space should always be added for GPT-2 as `" \n This"` is not tokenized the same as `" \nThis"` which GPT-2 may have seen more in its training data (unless white spaces are added after new lines as well)

### Future improvements

In the future, we should also add `prefix` and `suffix`. This could let us do things similar to GPT-3 with the inference widget and specific tasks:
* `padding_text` would be used to initialize tasks (ex: 10 examples conversation with a bot). This is always cropped out and never displayed in the output
* `prefix` would add required text (ex: `" Human: "`). This is displayed in output.
* `suffix` would add required text after prompt (ex: `" \n AI: "`). This is displayed in output.

borisdayma commented:
Another solution (though I think not as clean), would be to allow `prefix` element like in `TranslationPipeline`, though it is not as flexible (as it will become part of input and always display).

As a matter of fact, `TranslationPipeline`, `SummarizationPipeline` and `TextGenerationPipeline` are very similar (may benefit from some refactoring)!
As a temporary solution for my specific case, I could turn my GPT-2 into a translation model. Not sure it would trick the inference widget as I understand it detects first the model architecture to define which task to use… Also it's very hacky…

borisdayma commented:
I updated my [example](https://gist.github.com/borisdayma/89aaf1587390340add44c3c7081bffcd) of the feature with the case when it loads model config.

borisdayma commented:
Let me know if I can provide any additional information or make any changes

borisdayma commented:
Just checking if you need anything else from me @julien-c

julien-c commented:
I’m off for a couple of days so I’ll let others chime in

borisdayma commented:
Enjoy your time off!

borisdayma commented:
@julien-c @mfuntowicz sorry I'm just pinging.
I have a [demo](https://huggingface.co/huggingtweets/julien_c) I'm excited about and this feature would make it so much better.

julien-c commented:
Thoughts @patrickvonplaten?

julien-c commented:
Thoughts @patrickvonplaten @sshleifer? (text generation pipeline option)

sshleifer commented on examples/text-generation/run_generation.py:
```diff
@@ -180,7 +180,7 @@ def main():
     parser.add_argument("--p", type=float, default=0.9)
 
-    parser.add_argument("--padding_text", type=str, default="", help="Padding text for Transfo-XL and XLNet.")
+    parser.add_argument("--padding_text", type=str, default="", help="Padding text.")
```
why isn't this called prefix?

sshleifer replied:
Were we just ignoring this arg before?

borisdayma replied:
I agree that it should be renamed

sshleifer commented on src/transformers/pipelines.py:
```diff
@@ -675,27 +675,26 @@ def __call__(
+                padding = self._parse_and_tokenize(padding_text, padding=False, add_special_tokens=False)
+                # This impacts max_length and min_length argument that need adjusting.
+                padding_length = padding["input_ids"].shape[-1]
+                if "max_length" in generate_kwargs and generate_kwargs["max_length"] is not None:
```
```suggestion
                if generate_kwargs.get("max_length", None) is not None:
```

borisdayma replied:
Nice!

borisdayma commented:
This padding adds the script functionality to pipelines (and allow for custom text).

However the main idea would be to eventually allow "GPT-3 style" API's.

![image](https://user-images.githubusercontent.com/715491/91374590-2489fb80-e7de-11ea-8872-6f5ea8aeb2c7.png)

It could be renamed "prefix" but I suggest to make a distinction between padding and prefix where you would potentially want to display prefix & suffix in an interface (for example huggingface widget) but not necessarily the "padding".

This kind of API could be used for T5 and can let people create more modular inference examples.

borisdayma commented:
@sshleifer I gave more thoughts to it and I agree we should just name it `prefix` and maybe add a "deprecated warning" on `run_generation`.
It should be sufficient for now and we will always be able to add a more complex behavior in the future if necessary.

sshleifer commented:
style: you need to rebase and then `pip install black --upgrade`
Other CI failures are spurious, you can ignore them.

My last request would be to add test coverage for the new argument in test_pipelines.py.
see:
```
test_torch_text_generation
test_tf_text_generation
```

borisdayma commented:
Changes:

* renamed `padding_text` to `prefix` in pipelines and `run_generation.py` (with warning of deprecated argument)
* since PR was made, we cannot tokenize empty text anymore so we now check for empty "prefix"
* since PR was made, we cannot pass extra args to `generate` so we explicitly add `prefix` to pipeline `__call__` method
* added tests for pipeline using prefix

Note:

* we set default `prefix` to `None` (vs `""`) as we could pass `""` to overwrite model default value.

sshleifer commented on src/transformers/pipelines.py:
```diff
@@ -839,27 +847,27 @@ def __call__(
+                    prefix_length = prefix_inputs["input_ids"].shape[-1]
+                    if generate_kwargs.get("max_length", None) is not None:
+                        generate_kwargs["max_length"] += prefix_length
                     if "min_length" in generate_kwargs and generate_kwargs["min_length"] is not None:
```
same .get logic works on this line

sshleifer commented on src/transformers/pipelines.py:
```diff
@@ -839,27 +847,27 @@ def __call__(
-                    padding_text = self.PADDING_TEXT + self.tokenizer.eos_token
-                    padding = self._parse_and_tokenize(padding_text, padding=False, add_special_tokens=False)
+                    # For XLNet and TransformerXL we add an article to the prompt to give more state to the model.
+                    prefix = self.PREFIX
```
self.PREFIX -> self.XL_PREFIX if this is the only place it is used.

sgugger commented on src/transformers/pipelines.py:
```diff
@@ -821,6 +827,8 @@ def __call__(
                 Whether or not to include the decoded texts in the outputs.
             clean_up_tokenization_spaces (:obj:`bool`, `optional`, defaults to :obj:`False`):
                 Whether or not to clean up the potential extra spaces in the text output.
+            prefix (:obj:`str`, defaults to :obj:`None`):
```
```suggestion
            prefix (:obj:`str`, `optional`):
```
(Nit: no need to mention the defaults to None for something that is optional.)

borisdayma commented:
Thanks @LysandreJik. Yes I noticed later that my results were better without the initial space as well.

Intuitively it seemed to me like a waste of encoding differentiating between having a space or not (for example even after new lines) as the start of a document can simply be identified by the special token.
I thought there were more examples without the space within the original training set which would help my fine-tuning, as long as I do the same both for training and inference.
In the end, I get my best results with just `<|endoftext|>` between each sample and stripping every extra space.

borisdayma commented:
@LysandreJik let me know if anything else is required on my side.
Since it's a relatively large PR it can be hard to remain in sync with master.

patrickvonplaten commented:
I think we can merge this: @LysandreJik I leave it to you to merge :-)

julien-c commented:
+ @mfuntowicz can you update the transformers dependency in the inference API when this lands on master, so that @borisdayma can showcase his magic ;)

borisdayma commented:
@mfuntowicz can you let me know when you update the inference API so I do some tests on the widget?

mfuntowicz commented:
@borisdayma Everything has been deployed on the inference API, servers are up. 

Thanks for taking care of this 💪 🙏

borisdayma commented:
That was fast! Thanks it looks to be working great!
